source/main.cpp

```diff
@@ -91,313 +85,210 @@ int main(int argc, char *argv[])
+	}
+	catch(const runtime_error &error)
+	{
+		GameWindow::DoError(error.what());
+		return 1;
+	}
 		
-		// Check that the OpenGL version is high enough.
-		const char *glVersion = reinterpret_cast<const char *>(glGetString(GL_VERSION));
-		if(!glVersion || !*glVersion)
-			return DoError("Unable to query the OpenGL version!", window, context);
-		
-		const char *glslVersion = reinterpret_cast<const char *>(glGetString(GL_SHADING_LANGUAGE_VERSION));
-		if(!glslVersion || !*glslVersion)
+	// Remember the window state.
+	Preferences::Set("maximized", GameWindow::IsMaximized());		
+	Preferences::Set("fullscreen", GameWindow::IsFullscreen());
+	Screen::SetRaw(GameWindow::Width(), GameWindow::Height());
+	Preferences::Save();
+
+	GameWindow::Quit();
+	Audio::Quit();
+	
+	return 0;
+}
+
+void GameLoop(PlayerInfo &player, Conversation &conversation, bool &debugMode)
+{
+	// gamePanels is used for the main panel where you fly your spaceship. 
+	// All other game content related dialogs are placed on top of the gamePanels.
+	// If there are both menuPanels and gamePanels, then the menuPanels take
+	// priority over the gamePanels. The gamePanels will not be shown until
+	// the stack of menuPanels is empty.
+	UI gamePanels;
+	
+	// menuPanels is used for the panels related to pilot creation, preferences,
+	// game loading and game saving.
+	UI menuPanels;
+	
+	menuPanels.Push(new MenuPanel(player, gamePanels));
+	if(!conversation.IsEmpty())
+		menuPanels.Push(new ConversationPanel(player, conversation));
+
+	if(!GameWindow::HasSwizzle())
+		menuPanels.Push(new Dialog(
+			"Note: your computer does not support the \"texture swizzling\" OpenGL feature, "
+			"which Endless Sky uses to draw ships in different colors depending on which "
+			"government they belong to. So, all human ships will be the same color, which "
+			"may be confusing. Consider upgrading your graphics driver (or your OS)."));
+			
+	bool showCursor = true;
+	int cursorTime = 0;
+	int frameRate = 60; 
+	FrameTimer timer(frameRate);
+	bool isPaused = false;
+	
+	// If fast forwarding, keep track of whether the current frame should be drawn.
+	int skipFrame = 0;
+	
+	// Limit how quickly full-screen mode can be toggled.
+	int toggleTimeout = 0;
+	
+	// IsDone becomes true when the game is quit.
+	while(!menuPanels.IsDone())
+	{
+		if(toggleTimeout)
+			--toggleTimeout;
+			
+		// Handle any events that occurred in this frame.
+		SDL_Event event;
+		while(SDL_PollEvent(&event))
 		{
-			ostringstream out;
-			out << "Unable to query the GLSL version. OpenGL version is " << glVersion << ".";
-			return DoError(out.str(), window, context);
+			UI &activeUI = (menuPanels.IsEmpty() ? gamePanels : menuPanels);
+			
+			// If the mouse moves, reset the cursor movement timeout.
+			if(event.type == SDL_MOUSEMOTION)
+				cursorTime = 0;
+			
+			// The caps lock key slows the game down (to make it easier to
+			// see and debug things that are happening quickly).
```

Comment:
I realize this was another comment that you didn't write, and just moved, but it no longer applies to the block of code it introduces.

Reply:
Removed